Discovery service V1 to V2 upgrading guide (#1739)
neo4j/docs-operations: modules/ROOT/pages/clustering/setup/discovery.adoc

[[clustering-discovery]]
= Cluster server discovery

In order to join a running cluster, any new member must know the addresses of at least some of the other servers in the cluster.
This information is necessary to connect to the servers, run the discovery protocol, and obtain all the information about the cluster.

Neo4j provides several mechanisms for cluster members to discover each other and form a cluster based on the configuration and the environment in which the cluster is running, as well as the version of Neo4j being used.

[IMPORTANT]
====
From Neo4j 5.22, the current server discovery service has been marked as v1 and a new version v2 has been introduced.

> The new version v2 is recommended for new deployments.

Reviewer: I don't have a straightforward answer. There is a link on line 75 that points to an example of using LIST and the current v1. I think we need similar instructions on deploying a cluster with v2, but we can also do that later in a different PR.

[[clustering-discovery-list]]
=== Discovery using a list of server addresses

If the addresses of the other cluster members are known upfront, they can be listed explicitly.
However, this method has limitations, such as:

* If servers are replaced and the new members have different addresses, the list becomes outdated.
An outdated list can be avoided by ensuring that the new members can be reached via the same address as the old members, but this is not always practical.
* Under some circumstances the addresses are unknown when configuring the cluster.
This can be the case, for example, when using container orchestration to deploy a cluster.

To use this method, set `dbms.cluster.discovery.resolver_type=LIST` and hard code the addresses in the configuration of each server.
For example:

[.tabbed-example]
=====
[role=include-with-discovery-service-v1 label--deprecated-5.23]
======
[source, parameters]
----
dbms.cluster.discovery.resolver_type=LIST

server.discovery.advertised_address=server01.example.com:5000
dbms.cluster.discovery.endpoints=server01.example.com:5000,server02.example.com:5000,server03.example.com:5000
dbms.cluster.discovery.version=V1_ONLY
----
======

> [role=include-with-discovery-service-v2 label--new-5.23]

Author: I think we have this as `label--new-5.22` to keep consistent with the important notification above which says that v2 was introduced in 5.22

If this is not possible, consider using the discovery type `SRV`.
====

`dbms.cluster.discovery.resolver_type=SRV`::
With this configuration, the initial discovery members are resolved from _DNS SRV_ records to find the IP addresses/hostnames and discovery service ports (v1) *or* cluster advertised ports (v2) to contact.
+
The value of `dbms.cluster.discovery.endpoints` must be set to a single domain name and the port set to `0`.
The domain name returns a single SRV record when a DNS lookup is performed.
For example:
+
[.tabbed-example]
=====
[role=include-with-discovery-service-v1 label--deprecated-5.23]
======
[source, parameters]
----
dbms.cluster.discovery.resolver_type=SRV

server.discovery.advertised_address=server01.example.com:5000
dbms.cluster.discovery.endpoints=cluster01.example.com:0

dbms.cluster.discovery.version=V1_ONLY
----
======
[role=include-with-discovery-service-v2 label--new-5.23]
======
[source, parameters]
----
dbms.cluster.discovery.resolver_type=SRV

server.cluster.advertised_address=server01.example.com:6000
dbms.cluster.discovery.v2.endpoints=cluster01.example.com:0
dbms.cluster.discovery.version=V2_ONLY
----
======
=====
+

> The SRV record returned by DNS contains the IP address or hostname, and the cluster port for the servers to be discovered.

Author: So the original of this line is :
`
The SRV record returned by DNS should contain the IP address or hostname, and the discovery port for the servers to be discovered.
`

for v2 it'll be 
`
The SRV record returned by DNS should contain the IP address or hostname, and the **cluster** port for the servers to be discovered. 
`
So it might need to go into the tab or bold maybe? to make it clear of the difference

The discovery configuration is used for initial discovery and to continuously exchange information about changes to the topology of the cluster.

[[clustering-discovery-list]]
=== Discovery using a list of server addresses

If the addresses of the other cluster members are known upfront, they can be listed explicitly.
However, this method has limitations, such as:

* If servers are replaced and the new members have different addresses, the list becomes outdated.
An outdated list can be avoided by ensuring that the new members can be reached via the same address as the old members, but this is not always practical.
* Under some circumstances the addresses are unknown when configuring the cluster.
This can be the case, for example, when using container orchestration to deploy a cluster.

To use this method, set `dbms.cluster.discovery.resolver_type=LIST` and hard code the addresses in the configuration of each server.
For example:

[.tabbed-example]
=====

> [role=include-with-discovery-service-v1 label--deprecated-5.22]

Reviewer 2: Deprecated in 5.23, not 5.22. Right? 
line 17 says v1 is deprecated since 5.23

Reviewer: Yes, that's correct. v1 is deprecated from 5.23.

. Restart server01 with the new setting `dbms.cluster.discovery.version=V1_OVER_V2`.
+
image:discovery-service/in-place-1-v1-over-v2.png[]

. Then repeat this for server02 and server03.
Ensure to restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v1-over-v2.png[]

. Using `bolt://`, connect to servers 1, 2, 3, and run the following procedure.
It is important to connect via `bolt://` because otherwise the procedure might be routed and executed not on the intended server.
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+
They should display "Matching" in the `stateComparison`.
If they are not, wait and try again.
+
. Restart server1 again with the new setting `dbms.cluster.discovery.version=V2_OVER_V1`.
+
image:discovery-service/in-place-1-v2-over-v1.png[]

. Then repeat this for server2 and server03.
Ensure to restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v2-over-v1.png[]

. Do the sanity check similar to step 4.

. Repeat once more, restarting servers 1, 2, and 3 sequentially, with the new setting `dbms.cluster.discovery.version=V2_ONLY`
image:discovery-service/in-place-123-v2-only.png[]

> 

Reviewer: We can be more specific, yes.

When some members are running just one of the discovery services (V1 OR V2) and other members run both, the score stays permanently high.
This is no reason for worry.

. To fulfill this convergence, in different terminals, connect to server02 and server03 via `bolt://` and repeat steps 3 and 4 on both of them.

. Check the state on all servers again.
It should show that the states are 'Matching'.
+
[queryresult]
----
+--------------------------------+
| mode         | stateComparison |
+--------------------------------+
| "V1_OVER_V2" | "Matching"      |
+--------------------------------+
----

. On all three servers, run:
+
[source,cypher]
----
CALL dbms.cluster.switchDiscoveryServiceVersion("V2_OVER_V1");
----
+
At this point, v2 is the service that is running the cluster, with v1 running in the background.
+
[queryresult]
----
+--------------------------------+
| mode         | stateComparison |
+--------------------------------+
| "V2_OVER_V1" | "Matching"      |
+--------------------------------+
----

. Finally, turn off v1 by running the following procedure on all three servers:
+
[source,cypher]
----
CALL dbms.cluster.switchDiscoveryServiceVersion("V2_ONLY");
----

> . The `SHOW` procedure output now displays `V2_ONLY` running.

Reviewer: Good catch!

====

> As with DNS-based methods, the Kubernetes record lookup is only performed at startup.

Reviewer: It's actually mentioned in two other places:
line 34: _The discovery configuration is used for initial discovery and to continuously exchange information about changes to the topology of the cluster._
and right after the heading 
```
=== Discovery using DNS with multiple records
Where it is not practical or possible to explicitly list the addresses of cluster members to discover, you can use DNS-based mechanisms.
In such cases, a DNS record lookup is performed when a server starts up based on configuration settings.
Once a server has joined a cluster, further topology changes are communicated amongst the servers in the cluster as part of the discovery service.
```

> As with DNS-based methods, the Kubernetes record lookup is only performed at startup.

Reviewer: I'll update the Kubernetes section as well. Good catch, btw!

[[clustering-discovery-v1-to-v2]]
== Moving from discovery service v1 to v2

From Neo4j 5.23, you can move from the current discovery service v1 to the new version v2.

> This is a multi-stage process that might require some downtime. +

Reviewer: Yes, I agree with you and will remove this sentence.

It is important that both `dbms.cluster.discovery.endpoints` and `dbms.cluster.discovery.v2.endpoints` are set during the operation.
For more information, see <<clustering-discovery-dns>>.

* If `dbms.cluster.discovery.resolver_type=K8S`, set `dbms.kubernetes.discovery.v2.service_port_name` to the name of the service port used in the Kubernetes service definition for the Core's cluster port.
It is important that both `dbms.kubernetes.service_port_name` and `dbms.kubernetes.discovery.v2.service_port_name` are during the operation.
For more information, see <<clustering-discovery-k8s>>.

[[discovery-v1-to-v2-in-place]]
=== In-place rolling

[IMPORTANT]
====
In-place rolling reduces fault tolerance temporarily because you are restarting a running server.
To keep fault-tolerance, you can introduce a fourth server temporarily.
====
. Ensure that the new settings are added to the configuration as detailed in the preparation section.
. Restart server01 with the new setting `dbms.cluster.discovery.version=V1_OVER_V2`.
+
image:discovery-service/in-place-1-v1-over-v2.png[]

> . Then repeat this for server02 and server03.

Reviewer: ```suggestion
. Then repeat step 2 for server02 and server03.
```

Ensure to restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v1-over-v2.png[]

. Using `bolt://`, connect to servers 1, 2, 3, and run the following procedure.
It is important to connect via `bolt://` because otherwise the procedure might be routed and executed not on the intended server.
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+
They should display "Matching" in the `stateComparison`.
If they are not, wait and try again.
+
. Restart server1 again with the new setting `dbms.cluster.discovery.version=V2_OVER_V1`.
+
image:discovery-service/in-place-1-v2-over-v1.png[]

> . Then repeat this for server2 and server03.

Reviewer: ```suggestion
. Then repeat step 5 for server2 and server03.
```

Ensure to restart them sequentially, not in parallel.
+
image:discovery-service/in-place-23-v2-over-v1.png[]

> . Do the sanity check similar to step 4.

Reviewer: ```suggestion
. Similar to step 4, verify the `stateComparison` shows "Matching".
```

> . Repeat once more, restarting servers 1, 2, and 3 sequentially, with the new setting `dbms.cluster.discovery.version=V2_ONLY`

Reviewer: ```suggestion
. Repeat steps 5, 6 and 7, restarting servers 1, 2, and 3 sequentially, with the new setting `dbms.cluster.discovery.version=V2_ONLY`
```

image:discovery-service/in-place-123-v2-only.png[]

> . Verify that `CALL dbms.cluster.showParallelDiscoveryState()` now shows `V2_ONLY` running.

Reviewer 2: ```suggestion
. Verify that `CALL dbms.cluster.showParallelDiscoveryState()` now shows `V2_ONLY` running.

```

Reviewer: @NataliaIvakina, What is the suggestion here? I can't see any difference.

Reviewer 2: to add a line before the title

> .  Verify that `CALL dbms.cluster.showParallelDiscoveryState();` now shows `V2_ONLY` running.

Reviewer: ```suggestion
. Verify that `CALL dbms.cluster.showParallelDiscoveryState();` now shows `V2_ONLY` running.
```

[[discovery-v1-to-v2-new-server]]

> === New server rolling

Author: This title doesn't seem to be showing up in the doc: https://neo4j-docs-operations-1739.surge.sh/operations-manual/5/clustering/setup/discovery/

. Ensure that the new settings are added to the configuration as detailed in the preparation section.
. Start up three new servers with the setting `dbms.cluster.discovery.version=V1_OVER_V2`
+
image:discovery-service/v1_over_v2.png[]

. Using `bolt://`, connect to servers 4, 5, 6, and run the following procedure.
It is important to connect via `bolt://` because otherwise the procedure might be routed and executed not on the intended server.
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+

> The output should display "Matching" in the `stateComparison`.

Reviewer: ```suggestion
The output should display `Matching` in the `stateComparison`.
```

+
[source,cypher]
----
CALL dbms.cluster.switchDiscoveryServiceVersion("V1_OVER_V2");
----

. Check the state again:
+
[source,cypher]
----
CALL dbms.cluster.showParallelDiscoveryState();
----
+
Now the returned mode for this server must be `V1_OVER_V2` and the `stateComparison` must show that the states are not matching yet.
+
[queryresult]
----
+------------------------------------------------------------------------+
| mode         | stateComparison                                         |
+------------------------------------------------------------------------+
| "V1_OVER_V2" | "States are not matching after PT1M24.778S: (score:18)" |
+------------------------------------------------------------------------+
----
+
The score is a measure of how different the states are.
The score is 0 when the states are matching.
When some members are running just one of the discovery services (v1 or v2) and other members run both, the score stays permanently high.
This is no reason for worry.

. To fulfill this convergence, in different terminals, connect to server02 and server03 via `bolt://` and repeat steps 3 and 4 on both of them.

. Check the state on all servers again.

> It should show that the states are 'Matching'.

Reviewer: ```suggestion
It should show that the states are `Matching`.
```

+
[queryresult]
----
+--------------------------------+
| mode         | stateComparison |
+--------------------------------+
| "V1_OVER_V2" | "Matching"      |
+--------------------------------+
----

. On all three servers, run:
+
[source,cypher]
----
CALL dbms.cluster.switchDiscoveryServiceVersion("V2_OVER_V1");
----
+
At this point, v2 is the service that is running the cluster, with v1 running in the background.
+
[queryresult]
----
+--------------------------------+
| mode         | stateComparison |
+--------------------------------+
| "V2_OVER_V1" | "Matching"      |
+--------------------------------+
----

. Finally, turn off v1 by running the following procedure on all three servers:
+
[source,cypher]
----
CALL dbms.cluster.switchDiscoveryServiceVersion("V2_ONLY");
----

> . The `SHOW` procedure output now displays `V2_ONLY` running.

Reviewer: ```suggestion
.  Verify that `CALL dbms.cluster.showParallelDiscoveryState();` now shows `V2_ONLY` running.
```

+
[source,cypher]
----
CALL dbms.cluster.switchDiscoveryServiceVersion("V2_ONLY");
----

.  Verify that `CALL dbms.cluster.showParallelDiscoveryState();` now shows `V2_ONLY` running.
Note that `stateComparison` is `N/A` because you do not have v1 to compare states anymore.
+
[queryresult]
----
+-----------------------------+
| mode      | stateComparison |
+-----------------------------+
| "V2_ONLY" | "N/A"           |
+-----------------------------+
----
+
.Important
[IMPORTANT]
====
Remember to update the _neo4j.conf_ files for all the servers.
The switching using procedures does not persist anything to disk.
Therefore, when a server restarts, it starts right back with only v1 running.
As such, ensure that `dbms.cluster.discovery.version=V2_ONLY`, and that `dbms.cluster.discovery.v2.endpoints` or `dbms.kubernetes.discovery.v2.service_port_name`
are set as required, so that the servers start with v2 running on the next restart.
====

== Monitoring the move

When moving from the current discovery service v1 to the new version v2, you can monitor the progress using the following metrics:

* xref:monitoring/metrics/reference.adoc#discovery-service-V1[Discovery metrics v1]

> * xref:monitoring/metrics/reference.adoc#metrics-discovery-v2[Discovery metrics v2]

Author: Sorry, one more comment, I just noticed that we are not exposing `difference_score` in the metrics, only in the procedure. So I'm inclined for this part to be removed.

> Keep in mind that the difference score reported by the metrics and by the procedure does not always stay at 0.

Author: And to change this to `difference score report by the procedure`, i.e remove the metrics

Reviewer: Thanks, @yirensum. I'll change that. But should we still mention the metrics, or should I also remove those links?

Author: I think we should remove the mention of them, as the metrics dealing with different score are not public

Reviewer: I mean just the metrics without mentioning any scores. These metrics are public.

Author: I think it'll be nice to have reference to the metrics actually, so let's leave them here, and just remove mention of the difference score metric
